script.js

```diff
@@ -1,4 +1,165 @@
+  } else {
+    //condition 1 "if input == 1"
+    if (input == 1) {
+      playerFinalValue = Number(
+        `${currentPlayerRolls[0]}` + `${currentPlayerRolls[1]}`
+      );
+    }
+
+    //condition 2 "if input == 2"
+    if (input == 2) {
+      playerFinalValue = Number(
+        `${currentPlayerRolls[1]}` + `${currentPlayerRolls[0]}`
+      );
+    }
+    allPlayersScore.push(playerFinalValue); // add the final values to the corresponding array on line 19
+    currentPlayerRolls = []; //clear the current player rolls array
+
+    return (
+      "Player " +
+      currentPlayer +
+      ", your final value is " +
+      playerFinalValue +
+      "."
+    );
+  }
+};
+var comparePlayersScore = function () {
+  //player 1 wins:
+  if (allPlayersScore[0] > allPlayersScore[1]) {
+    return (
+      "Player's 1 final value is " +
+      allPlayersScore[0] +
+      ". <br> <br> Player's 2 final value is " +
+      allPlayersScore[1] +
+      ". <br> <br> " +
+      "Player 1 wins."
+    );
+    //player 2 wins
+  } else if (allPlayersScore[1] > allPlayersScore[0]) {
+    return (
+      "Player's 1 final value is " +
+      allPlayersScore[0] +
+      ". <br> <br> Player's 2 final value is " +
+      allPlayersScore[1] +
+      ". <br> <br> " +
+      "Player 2 wins."
+    );
+    // a tie:
+  } else {
+    return (
+      "Player's 1 final value is " +
+      allPlayersScore[0] +
+      ". <br> <br> Player's 2 final value is " +
+      allPlayersScore[1] +
+      ". <br> <br> " +
+      "It's a tie."
+    );
+  }
+};
+//main function
 var main = function (input) {
-  var myOutputValue = 'hello world';
-  return myOutputValue;
+  var myOutputMessage = "";
+  if (gameState == GAME_STATE_ROLL_DICE) {
+    myOutputMessage = rollDiceForPlayer(); // call the roll dice x2 function
+    //cnange the game state
+    gameState = GAME_STATE_CHOOSE_DICE_ORDER;
+
+    return myOutputMessage;
+  }
+  if (gameState == GAME_STATE_CHOOSE_DICE_ORDER) {
+    myOutputMessage = getPlayerFinalValue(input); //output the player value
+    //if the input is incorrect, call the function again:
+    if (myOutputMessage == errorMessage) {
+      myOutputMessage = getPlayerFinalValue(input);
+      return myOutputMessage;
+    }
+
+    if (currentPlayer == 1) {
+      currentPlayer = 2;
+      gameState = GAME_STATE_ROLL_DICE;
+
+      return myOutputMessage + "<br> <br> It is now player's 2 turn.";
+    }
+
+    if (currentPlayer == 2) {
+      gameState = GAME_STATE_COMPARE_SCORES;
+
+      return (
+        myOutputMessage +
+        "<br> <br> Please click 'submit' to see the final score."
+      );
+    }
+  }
+  if (gameState == GAME_STATE_COMPARE_SCORES) {
+    myOutputMessage = comparePlayersScore();
+    allPlayersScore = [];
+
+    gameState = GAME_STATE_PLAY_AGAIN;
+    return myOutputMessage;
+  }
+  if (gameState == GAME_STATE_PLAY_AGAIN) {
+    gameState = GAME_STATE_ROLL_DICE;
+    currentPlayer = 1;
+    myOutputMessage = main();
+  }
```

Comment:
Interesting, this is very good understanding of control flow. However, i wouldn't recommend this pattern because it is unnecessary complexity. We have 4 game modes, and within each we have a `if` checking and `return` within that `if`. That means that if code manages to reach this point , `gameState == GAME_STATE_PLAY_AGAIN` is implicitly true already.

Maybe we just need a default `myOutputMessage` and there would be no need for `myOutputMessage = main()`. Well done regardless! happy to see that your control flow is top tier now